prov/verbs/src/windows/verbs_nd_ov.c

```diff
@@ -0,0 +1,399 @@
+		       FI_LOG_EP_CTRL,
+		       "IND2Connector::GetPrivateData: hr=0x%08lx\n", hr);
+	} else {
+		entry_nd->event.param.conn.private_data = NULL;
+	}
+}
+
+// Used in conjunction with IND2Listener::GetConnectionRequest()
+void vrb_nd_cm_connreq_event(vrb_nd_event_base *base, DWORD bytes)
+{
+	trace(__func__);
+
+	struct rdma_cm_event_vrb_nd *entry_nd =
+		(struct rdma_cm_event_vrb_nd *)calloc(
+			1, sizeof(struct rdma_cm_event_vrb_nd));
+	assert(entry_nd);
+
+	vrb_nd_cm_listen_event *event =
+		container_of(base, vrb_nd_cm_listen_event, base);
+	IND2Connector *connector = event->connector;
+
+	struct rdma_cm_id *id;
+	int ret = rdma_create_id(event->listen_id->channel, &id,
+				 event->listen_id->context, RDMA_PS_TCP);
+	assert(!ret);
+	id->verbs = event->listen_id->verbs;
+	struct rdma_cm_id_vrb_nd *id_nd =
+		container_of(id, struct rdma_cm_id_vrb_nd, id);
+	struct rdma_cm_id_vrb_nd *listen_id_nd =
+		container_of(event->listen_id, struct rdma_cm_id_vrb_nd, id);
+	id_nd->connector = listen_id_nd->connector;
+
+	entry_nd->event.id = id;
+	entry_nd->event.listen_id = event->listen_id;
+	entry_nd->event.event = RDMA_CM_EVENT_CONNECT_REQUEST;
+	entry_nd->event.status = 0;
+
+	vrb_nd_get_read_limits(connector, entry_nd);
+	vrb_nd_get_connection_data(connector, entry_nd);
+
+	ULONG len = sizeof(id->route.addr.src_addr);
+	HRESULT hr = connector->lpVtbl->GetLocalAddress(
+		connector, &id->route.addr.src_addr, &len);
+	FI_LOG(&vrb_prov, FAILED(hr) ? FI_LOG_WARN : FI_LOG_DEBUG,
+	       FI_LOG_EP_CTRL, "IND2Connector::GetLocalAddress: hr=0x%08lx\n",
+	       hr);
+
+	len = sizeof(id->route.addr.dst_addr);
+	hr = connector->lpVtbl->GetPeerAddress(connector,
+					       &id->route.addr.dst_addr, &len);
+	FI_LOG(&vrb_prov, FAILED(hr) ? FI_LOG_WARN : FI_LOG_DEBUG,
+	       FI_LOG_EP_CTRL, "IND2Connector::GetPeerAddress: hr=0x%08lx\n",
+	       hr);
+
+	struct rdma_event_channel_vrb_nd *ch_nd = container_of(
+		id->channel, struct rdma_event_channel_vrb_nd, channel);
+	dlistfd_insert_tail(&entry_nd->entry, &ch_nd->q);
+	VRB_DBG(FI_LOG_EQ, "EQ SET event:%p %p %d\n", entry_nd->event.id,
+		entry_nd->event.listen_id, entry_nd->event.event);
+
+	// Need to obtain a new connector for the listening id.
+	ret = rdma_bind_addr(&listen_id_nd->id,
+			     &listen_id_nd->id.route.addr.src_addr);
+	assert(!ret);
+
+	vrb_nd_cm_listen_event *conn = (vrb_nd_cm_listen_event *)calloc(
+		1, sizeof(vrb_nd_cm_listen_event));
+	assert(conn);
+
+	conn->base.event_cb = vrb_nd_cm_connreq_event;
+	conn->base.error_cb = vrb_nd_cm_connreq_error;
+	conn->connector = listen_id_nd->connector;
+	conn->listen_id = &listen_id_nd->id;
+
+	hr = listen_id_nd->listener->lpVtbl->GetConnectionRequest(
+		listen_id_nd->listener, (IUnknown *)listen_id_nd->connector,
+		&conn->base.ov);
+	FI_LOG(&vrb_prov, FAILED(hr) ? FI_LOG_WARN : FI_LOG_DEBUG,
+	       FI_LOG_EP_CTRL,
+	       "IND2Listener::GetConnectionRequest: hr=0x%08lx; ov=%p\n", hr,
+	       &conn->base.ov);
+
+	free(base);
```

Comment:
see prior comment

Reply:
So this was the pattern that netdir was using, which I largely left intact, to deal with asynchronous overlapped requests and completions.

The requestor would allocate a buffer with an OVERLAPPED structure at the beginning, followed by functions to call for success/failure, and then any additional data that was needed by the completion routine to process things.

All completions were routed to a common callback function, nd_io_cb, then handled from there as needed.

Upon completion, the completion routine would free the buffer that had been used.

Would you be alright with just adding comments at the allocation and free locations about the usage pattern or would you prefer some sort of structural changes?

Reply:
I'd prefer some sort of structural change so that the lifetime of the objects is clear.  Let's assume whatever the current netdir provider is doing is likely wrong and unmaintainable.  :)